array.h

```diff
@@ -808,6 +811,16 @@ T convert_dims(const U& u, std::index_sequence<Is...>) {
   return std::make_tuple(convert_dim<Is, T>(u)...);
 }
 
+constexpr index_t factorial(index_t x) {
+  return x == 1 ? 1 : x * factorial(x - 1);
+}
+
+// The errors that result from not satisfying this check are probably hell,
```

Comment:
factorial isn't enough to guarantee that it's a permutation right? You may have repeated elements that just happen to multiply. Forcing the two ranks be the same helps a bit.

In C++20, std::sort is constexpr. It looks possible in C++17: https://medium.com/@vgasparyan1995/compile-time-merge-sort-c-bb0ace62cc23

Then we can sort the Is..., and then check that it's equal to the integer_sequence?

Reply:
I'd rather stick to C++14 if possible, plus I'm a bit wary of compile times. I agree checking the length of Is... would help. I think that if we require sizeof...(Is) == Rank, that the factorial check is reliable, or at least I can't think of a counterexample. I'll add that.

Reply:
I think it'll only be a problem when the rank is large. I'm too lazy to figure out exactly the case since 1*2*3*4 is already 24. I was imagining multiplying one of the entries by 2 and dividing another by 2 - the product will be the same.